File: app/src/main/java/org/beiwe/app/listeners/AmbientAudioListener.java
package org.beiwe.app.listeners;

import static org.beiwe.app.UtilsKt.printe;
import static org.beiwe.app.UtilsKt.printi;

import android.content.Context;
import android.media.MediaRecorder;
import android.os.AsyncTask;
import android.util.Log;

import org.beiwe.app.MainService;
import org.beiwe.app.R;
import org.beiwe.app.Timer;
import org.beiwe.app.storage.AudioFileManager;
import org.beiwe.app.storage.PersistentData;
import org.beiwe.app.storage.TextFileManager;

import java.io.File;
import java.io.IOException;

/**
 * AmbientAudioListener is a Singleton class: it should only be started and instantiated once.
 */
public class AmbientAudioListener {
    private static Context appContext;
    private static MediaRecorder mRecorder;
    private static final String filenameExtension = ".mp4";
    public static final String unencryptedTempAudioFilename = "tempUnencryptedAmbientAudioFile";
    public static String currentlyBeingWrittenEncryptedFilename = null;
    private static int deviceSetupCount = 0;
    private static boolean instantiated = false;

    private static String getUnencryptedAudioFilepath() {
        return appContext.getFilesDir().getAbsolutePath() + "/" + unencryptedTempAudioFilename;
    }

    public static boolean isCurrentlyRunning() {
        // if both class variables are instantiated then we SHOULD be recording, have to watch
        // the audio file size to confirm.
        return (instantiated && mRecorder != null);
    }
    
    public static synchronized void startRecording(Context applicationContext) {
        TextFileManager.writeDebugLogStatement("AmbientAudioListener.startRecording()");
        if (!instantiated) {
            // first run
            // Instantiate the AmbientAudioRecorder only if it has not yet been instantiated
//            ambientAudioListenerInstance = new AmbientAudioListener();
            appContext = applicationContext;
            
            // if there is an extant temp audio file and the ambient recording feature is just
            // starting up then we need to encrypt the existing file
            File temp_audio_file = applicationContext.getFileStreamPath(unencryptedTempAudioFilename);
            if (temp_audio_file.exists() && currentlyBeingWrittenEncryptedFilename == null) {
                forceEncrypt();  // for now we don't care if this stalls the background service
            }
            instantiated = true;
        }
        
        // Start the Media Recorder only if it is not currently running
        if (mRecorder == null) {
            try {
                deviceSetupCount += 1;
                mRecorder = new MediaRecorder();
                mRecorder.reset();
                mRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
                mRecorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
                mRecorder.setOutputFile(getUnencryptedAudioFilepath());
                mRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.HE_AAC);
                mRecorder.setAudioChannels(1);
                mRecorder.setAudioSamplingRate(((int) PersistentData.getAmbientAudioSampleRate()));
                mRecorder.setAudioEncodingBitRate((int) PersistentData.getAmbientAudioBitrate());
                TextFileManager.writeDebugLogStatement("AmbientAudioListener device setup success after " + deviceSetupCount + " attempt(s)");
            } catch (java.lang.RuntimeException e) {
                // TODO: does this occur only when certain security details (device lock) are active, or is it truly manufacturer-based
                // this causes a lot of spam when the screen is off and the microphone is being turned on.
                printe("AmbientAudioListener", "AmbientAudioListener device setup failed, count is at " + deviceSetupCount);
                e.printStackTrace();
                if (deviceSetupCount == 1) {
                    TextFileManager.writeDebugLogStatement("AmbientAudioListener device setup failed");
                    TextFileManager.writeDebugLogStatement(e.getMessage());
                }
                mRecorder = null;
                return;
            }
            deviceSetupCount = 0;
            
            try {
                mRecorder.prepare();
            } catch (IOException e) {
                Log.e("AmbientAudioListener", "MediaRecorder.prepare() failed");
                e.printStackTrace();
                TextFileManager.writeDebugLogStatement("AmbientAudioListener MediaRecorder.prepare() failed");
                TextFileManager.writeDebugLogStatement(e.getMessage());
                return;
            }
            mRecorder.start();
        }
    }
    
    public static synchronized void encryptAmbientAudioFile() {
        TextFileManager.writeDebugLogStatement("AmbientAudioListener.encryptAmbientAudioFile()");
        if (instantiated && mRecorder != null) {
            // If the audio recorder exists, stop recording and start encrypting the file
            mRecorder.stop();
            mRecorder.reset();
            mRecorder.release();
            new EncryptAmbientAudioFileTask().execute();
        }
    }
    
    private static void forceEncrypt(){
        // we need an encrypt capacity in case the app starts and sees an unencrypted audio file.
        currentlyBeingWrittenEncryptedFilename = AudioFileManager.generateNewEncryptedAudioFileName(null, filenameExtension);
        AudioFileManager.encryptAudioFile(getUnencryptedAudioFilepath(), currentlyBeingWrittenEncryptedFilename, appContext);
        AudioFileManager.delete(unencryptedTempAudioFilename);
        currentlyBeingWrittenEncryptedFilename = null;
        TextFileManager.writeDebugLogStatement("recovered potentially lost or corrupted audio file, timestamp will be inaccurate: " + currentlyBeingWrittenEncryptedFilename);
    }
    
    private static class EncryptAmbientAudioFileTask extends AsyncTask<Void, Void, Void> {
        @Override
        protected void onPreExecute() {
            // Before doing anything else, set the filename of the encrypted file. This tells
            // TextFileManager.getAllUploadableFiles NOT to upload it until it's finished writing.
            currentlyBeingWrittenEncryptedFilename = AudioFileManager.generateNewEncryptedAudioFileName(null, filenameExtension);
        }

        @Override
        protected Void doInBackground(Void... params) {
            AudioFileManager.encryptAudioFile(getUnencryptedAudioFilepath(), currentlyBeingWrittenEncryptedFilename, appContext);
            return null;
        }

        @Override
        protected void onPostExecute(Void arg) {
            // Clear the filename, so TextFileManager.getAllUploadableFiles can now upload it
            currentlyBeingWrittenEncryptedFilename = null;
            // Set the Media Recorder back to null to free it up to run again
            mRecorder = null;
            // Delete the unencrypted temp audio file
            AudioFileManager.delete(unencryptedTempAudioFilename);
            // Restart recording
            startRecording(appContext);
        }
    }
}

File: app/src/main/java/org/beiwe/app/listeners/AmbientAudioListener.kt
package org.beiwe.app.listeners

import android.content.Context
import android.media.MediaRecorder
import android.os.AsyncTask
import android.util.Log
import org.beiwe.app.printe
import org.beiwe.app.storage.AudioFileManager
import org.beiwe.app.storage.AudioFileManager.generateAmbientEncryptedAudioFileName
import org.beiwe.app.storage.PersistentData
import org.beiwe.app.storage.TextFileManager
import java.io.IOException

/**
 * AmbientAudioListener is a !Singleton class: it is only instantiated once.
 * The pattern followed here is not in line with the rest of the codebase's listeners, but
 * might be the more correct pattern avoiding application context leaks, which I don't understand anyway.
 */

const val filenameExtension = ".mp4"
const val ambientTempAudioFilename = "tempUnencryptedAmbientAudioFile"
const val FILE_ERROR = "recovered potentially lost or corrupted audio file, timestamp will be inaccurate: "
const val AUDIO_FAIL_COUNT = "AmbientAudioListener device setup failed, count is at "
const val AMBIENT_SETUP_SUCCESS = "AmbientAudioListener device setup success after "

object AmbientAudioListener {
    private var mRecorder: MediaRecorder? = null
    var appContext: Context? = null

    @JvmField
    var currentlyWritingEncryptedFilename: String? = null
    private var deviceSetupCount = 0
    private var instantiated = false

    // if both class variables are instantiated then we SHOULD be recording, would have to watch
    // the audio file size to confirm I guess?
    val isCurrentlyRunning: Boolean
        get() = instantiated && mRecorder != null

    val unencryptedAudioFilepath: String
        get() = appContext!!.filesDir.absolutePath + "/" + ambientTempAudioFilename

    val ambient_audio_off_action = {
        encryptAmbientAudioFile()
    }

    @Synchronized
    fun startRecording(applicationContext: Context) {
        TextFileManager.writeDebugLogStatement("AmbientAudioListener.startRecording()")

        if (!instantiated) {  // first run
            appContext = applicationContext
            // if there is an extant temp audio file and the ambient recording feature is just
            // starting up then we need to encrypt the existing file.
            val temp_audio_file = applicationContext.getFileStreamPath(ambientTempAudioFilename)
            if (temp_audio_file.exists() && currentlyWritingEncryptedFilename == null) {
                forceEncrypt() // for now we don't care if this stalls the background service for a second
            }
            instantiated = true
        }

        // Start the Media Recorder only if it is not currently running
        if (mRecorder == null) {
            try {
                deviceSetupCount += 1
                mRecorder = MediaRecorder(appContext!!)
                mRecorder!!.reset()
                mRecorder!!.setAudioSource(MediaRecorder.AudioSource.MIC)
                mRecorder!!.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
                mRecorder!!.setOutputFile(unencryptedAudioFilepath)
                mRecorder!!.setAudioEncoder(MediaRecorder.AudioEncoder.HE_AAC)
                mRecorder!!.setAudioChannels(1)
                mRecorder!!.setAudioSamplingRate(PersistentData.getAmbientAudioSampleRate().toInt())
                mRecorder!!.setAudioEncodingBitRate(PersistentData.getAmbientAudioBitrate().toInt())
                TextFileManager.writeDebugLogStatement(AMBIENT_SETUP_SUCCESS + deviceSetupCount + " attempt(s)")
            } catch (e: RuntimeException) {
                // TODO: does this occur only when certain security details (device lock) are active,
                //  or is it truly manufacturer-based.
                // this causes a lot of spam when the screen is off and the microphone is being turned on.
                printe("AmbientAudioListener", AUDIO_FAIL_COUNT + deviceSetupCount)
                e.printStackTrace()
                if (deviceSetupCount == 1) {
                    TextFileManager.writeDebugLogStatement("AmbientAudioListener device setup failed")
                    TextFileManager.writeDebugLogStatement(e.message)
                }
                mRecorder = null
                return
            }

            deviceSetupCount = 0  // reset setup attempt count

            try {
                mRecorder!!.prepare()
            } catch (e: IOException) {
                Log.e("AmbientAudioListener", "MediaRecorder.prepare() failed")
                e.printStackTrace()
                TextFileManager.writeDebugLogStatement("AmbientAudioListener MediaRecorder.prepare() failed")
                TextFileManager.writeDebugLogStatement(e.message)
                return
            }
            mRecorder!!.start()  // ok now actually start (never seen to fail)
        }
    }

    @Synchronized
    fun encryptAmbientAudioFile() {
        TextFileManager.writeDebugLogStatement("AmbientAudioListener.encryptAmbientAudioFile()")
        // If the audio recorder exists, stop recording and start encrypting the file
        if (instantiated && mRecorder != null) {
            mRecorder!!.stop()
            mRecorder!!.reset()
            mRecorder!!.release()
            EncryptAmbientAudioFileTask().execute()  // executes on a thread and then starts a new recording
        }
    }

    private fun forceEncrypt() {
        // we need an encrypt capacity in case the app starts and sees an unencrypted audio file.
        currentlyWritingEncryptedFilename = generateAmbientEncryptedAudioFileName(filenameExtension)
        AudioFileManager.encryptAudioFile(
                unencryptedAudioFilepath,
                currentlyWritingEncryptedFilename,
                appContext
        )
        AudioFileManager.delete(ambientTempAudioFilename)
        currentlyWritingEncryptedFilename = null
        TextFileManager.writeDebugLogStatement(FILE_ERROR + currentlyWritingEncryptedFilename)
    }

    //TODO: make this a job?
    private class EncryptAmbientAudioFileTask : AsyncTask<Void?, Void?, Void?>() {
        override fun onPreExecute() {
            // Before doing anything else, set the filename of the encrypted file. This tells
            // TextFileManager.getAllUploadableFiles NOT to upload it until it's finished writing.
            // TODO: currently using null for special behavior; build the special behavior.
            currentlyWritingEncryptedFilename = generateAmbientEncryptedAudioFileName(filenameExtension)
        }

        override fun doInBackground(vararg params: Void?): Void? {
            AudioFileManager.encryptAudioFile(
                    unencryptedAudioFilepath,
                    currentlyWritingEncryptedFilename,
                    appContext
            )
            return null
        }

        override fun onPostExecute(arg: Void?) {
            // Clear the filename, so TextFileManager.getAllUploadableFiles can now upload it
            currentlyWritingEncryptedFilename = null
            // Set the Media Recorder back to null to free it up to run again (apparently necessary?)
            mRecorder = null
            // Delete the unencrypted temp audio file, Restart recording
            AudioFileManager.delete(ambientTempAudioFilename)
            startRecording(appContext!!)
        }
    }
}